### PR TITLE
GAP-2138: Remove an editor endpoint

### PR DESCRIPTION
**Pull request by dominicwest**
## Description

Ticket [GAP-2138](https://technologyprogramme.atlassian.net/jira/software/projects/GAP/boards/216?selectedIssue=GAP-2138)

Summary of the changes and the related issue. List any dependencies that are required for this change:
- Remove editor endpoint
- Throws 403 if editor = createdBy column of scheme
- Throws 404 if scheme or grant not found

## Type of change

Please check the relevant options.

- [ ] Bug fix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] This change requires a documentation update.

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes:

- [x] Unit Test

- [ ] Integration Test (if applicable)

- [ ] End-to-End Test (if applicable)

## Screenshots (if appropriate):

Please attach screenshots of the change if it is a UI change:

Screenshot 2024-03-01 at 16 13 08

https://github.com/cabinetoffice/gap-find-apply-web/assets/101722961/fb678361-86e9-4113-b0c7-d42481dbb116

# Checklist:

- [x] If I have listed dependencies above, I have ensured that they are present in the target branch.
- [x] I have performed a self-review of my code.
- [x] I have commented my code in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation where applicable.
- [ ] I have run cypress tests, and they all pass locally.

[GAP-2138]: https://technologyprogramme.atlassian.net/browse/GAP-2138?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/main/java/gov/cabinetoffice/gap/adminbackend/controllers/SchemeEditorController.java [hunk 1]
```diff
@@ -59,4 +59,10 @@ public ResponseEntity<String> addEditorToScheme(@PathVariable final Integer sche
         return ResponseEntity.ok("Editor added to scheme successfully");
     }
 
+    @DeleteMapping(value = "/{editorId}")
+    public ResponseEntity<String> deleteSchemeEditor(@PathVariable Integer schemeId, @PathVariable Integer editorId) {
```
Comment (dylanwrightCO) [inline src/main/java/gov/cabinetoffice/gap/adminbackend/controllers/SchemeEditorController.java:63]: Can we add an ownership check to this endpoint?

File: src/main/java/gov/cabinetoffice/gap/adminbackend/services/SchemeEditorService.java [hunk 2]
```diff
@@ -111,4 +113,17 @@ public SchemeEntity addEditorToScheme(final Integer schemeId, final String edito
...
+        final SchemeEntity scheme = schemeRepo.findById(schemeId)
+                .orElseThrow(() -> new NotFoundException("Delete scheme editor: Scheme not found"));
+
+        if (scheme.getCreatedBy().equals(editorId))
+            throw new ForbiddenException("Delete scheme editor: Cannot delete scheme creator");
+
+        final GrantAdmin grantAdmin = grantAdminRepository.findById(editorId)
+                .orElseThrow(() -> new NotFoundException("Delete scheme editor: Grant Admin with editor access not found for this scheme"));
+
+        scheme.removeAdmin(grantAdmin);
```
Comment (dylanwrightCO) [inline src/main/java/gov/cabinetoffice/gap/adminbackend/services/SchemeEditorService.java:126]: What happens if the admin exists but is not an editor on the scheme?

File: src/main/java/gov/cabinetoffice/gap/adminbackend/controllers/SchemeEditorController.java (hunk 1 above)
Comment (dominicwest) [reply src/main/java/gov/cabinetoffice/gap/adminbackend/controllers/SchemeEditorController.java:63]: We only want scheme owners to be able to hit this endpoint

File: src/main/java/gov/cabinetoffice/gap/adminbackend/services/SchemeEditorService.java (hunk 2 above)
Comment (dominicwest) [reply src/main/java/gov/cabinetoffice/gap/adminbackend/services/SchemeEditorService.java:126]: As per check above this - admin must own this scheme. Editors should not be able to remove other editors as per the ACs

**Review by dylanwrightCO — APPROVED** [review]

**Pull request merged**